File: floors/haskell/DepartmentFloor.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}

{-|
Module      : DepartmentFloor
Description : Haskell Department Floor - Floor 20
Copyright   : (c) Miniature Office, 2024
License     : MIT

FLOOR 20 - HASKELL JURISDICTION
Domain: Pure functional programming, Type-safe systems, Compilers
Architectural Law: Type safety > runtime checks, No partial functions, Monadic error handling
-}

module Main (main) where

import Control.Monad (forever)
import Control.Monad.IO.Class (liftIO)
import Control.Monad.Trans.State.Strict (StateT, evalStateT, get, modify)
import Data.Aeson (ToJSON(..), FromJSON(..), Value(..), object, (.=), (.:), (.:?))
import qualified Data.Aeson as JSON
import qualified Data.ByteString.Lazy.Char8 as BSL
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import qualified Data.Text.IO as TIO
import GHC.Generics (Generic)
import System.IO (hFlush, hPutStrLn, stdout, stderr)

import FloorAgent (FloorAgent, AgentId, Role, Capability, createAgent)
import TaskManager (Task, TaskState, emptyTaskState, createTask, getAllTasks)
import CodeAnalyzer (analyzeCode, formatCode, lintCode)


-- Floor State
data FloorState = FloorState
    { floorAgents :: Map AgentId FloorAgent
    , floorTasks :: TaskState
    } deriving (Show, Generic)

emptyFloorState :: FloorState
emptyFloorState = FloorState Map.empty emptyTaskState

type FloorM = StateT FloorState IO

-- JSON-RPC Request
data RPCRequest = RPCRequest
    { rpcMethod :: Text
    , rpcParams :: Maybe Value
    } deriving (Show, Generic)

instance FromJSON RPCRequest where
    parseJSON = JSON.withObject "RPCRequest" $ \v -> RPCRequest
        <$> v .: "method"
        <*> v .:? "params"

-- JSON-RPC Response
data RPCResponse = RPCResponse
    { responseData :: Value
    } deriving (Show, Generic)

instance ToJSON RPCResponse where
    toJSON (RPCResponse val) = val

-- Floor Info
data FloorInfo = FloorInfo
    { floorNumber :: Int
    , floorLanguage :: Text
    , domain :: Text
    , architecturalLaw :: Text
    , securityDoctrine :: Text
    , offices :: [Text]
    , agentCount :: Int
    , taskCount :: Int
    , agents :: [FloorAgent]
    , tasks :: [Task]
    } deriving (Show, Generic)

instance ToJSON FloorInfo where
    toJSON info = object
        [ "floor_number" .= floorNumber info
        , "language" .= floorLanguage info
        , "domain" .= domain info
        , "architectural_law" .= architecturalLaw info
        , "security_doctrine" .= securityDoctrine info
        , "offices" .= offices info
        , "agent_count" .= agentCount info
        , "task_count" .= taskCount info
        , "agents" .= agents info
        , "tasks" .= tasks info
        ]

-- Request Parameters
data AddAgentParams = AddAgentParams
    { paramAgentId :: Text
    , paramName :: Text
    , paramRole :: Role
    , paramCapabilities :: [Capability]
    } deriving (Show, Generic)

instance FromJSON AddAgentParams where
    parseJSON = JSON.withObject "AddAgentParams" $ \v -> AddAgentParams
        <$> v .: "agent_id"
        <*> v .: "name"
        <*> v .: "role"
        <*> v .: "capabilities"

data CreateTaskParams = CreateTaskParams
    { paramTaskId :: Text
    , paramTitle :: Text
    , paramAssignedTo :: Text
    } deriving (Show, Generic)

instance FromJSON CreateTaskParams where
    parseJSON = JSON.withObject "CreateTaskParams" $ \v -> CreateTaskParams
        <$> v .: "task_id"
        <*> v .: "title"
        <*> v .: "assigned_to"

data ProcessCodeParams = ProcessCodeParams
    { paramCode :: Text
    , paramOperation :: Text
    } deriving (Show, Generic)

instance FromJSON ProcessCodeParams where
    parseJSON = JSON.withObject "ProcessCodeParams" $ \v -> ProcessCodeParams
        <$> v .: "code"
        <*> v .: "operation"

-- Handlers
handleGetInfo :: FloorM Value
handleGetInfo = do
    state <- get
    let agentList = Map.elems (floorAgents state)
        taskList = getAllTasks (floorTasks state)
        info = FloorInfo
            { floorNumber = 20
            , floorLanguage = "haskell"
            , domain = "Pure functional programming, Type-safe systems, Compilers"
            , architecturalLaw = "Type safety > runtime checks, No partial functions, Monadic error handling"
            , securityDoctrine = "Type-level security, Safe API design, No unsafe operations"
            , offices = ["Architecture Office", "Implementation Office", "Review Office", 
                         "Test Office", "Security Office", "Manager Office"]
            , agentCount = Map.size (floorAgents state)
            , taskCount = length taskList
            , agents = agentList
            , tasks = taskList
            }
    return $ toJSON info

handleAddAgent :: Value -> FloorM Value
handleAddAgent params = do
    case JSON.fromJSON params of
        JSON.Error err -> return $ object ["status" .= ("error" :: Text), "message" .= err]
        JSON.Success (AddAgentParams aid name role caps) -> do
            let agent = createAgent aid name role caps
            modify $ \s -> s { floorAgents = Map.insert aid agent (floorAgents s) }
            return $ object ["status" .= ("success" :: Text), "agent" .= agent]

handleCreateTask :: Value -> FloorM Value
handleCreateTask params = do
    case JSON.fromJSON params of
        JSON.Error err -> return $ object ["status" .= ("error" :: Text), "message" .= err]
        JSON.Success (CreateTaskParams tid title assignee) -> do
            task <- createTask tid title assignee
            modify $ \s -> s { floorTasks = Map.insert tid task (floorTasks s) }
            return $ object ["status" .= ("success" :: Text), "task" .= task]

handleProcessCode :: Value -> FloorM Value
handleProcessCode params = do
    case JSON.fromJSON params of
        JSON.Error err -> return $ object ["status" .= ("error" :: Text), "message" .= err]
        JSON.Success (ProcessCodeParams code op) ->
            case op of
                "analyze" -> do
                    let analysis = analyzeCode code
                    return $ object ["status" .= ("success" :: Text), "analysis" .= analysis]
                "format" ->
                    case formatCode code of
                        Left err -> return $ object ["status" .= ("error" :: Text), "message" .= err]
                        Right formatted -> return $ object
                            [ "status" .= ("success" :: Text)
                            , "formatted" .= True
                            , "code" .= formatted
                            ]
                "lint" ->
                    case lintCode code of
                        Left err -> return $ object ["status" .= ("error" :: Text), "message" .= err]
                        Right warnings -> return $ object
                            [ "status" .= ("success" :: Text)
                            , "warnings" .= warnings
                            ]
                _ -> return $ object ["status" .= ("error" :: Text), "message" .= ("Unknown operation: " <> op)]

handleRequest :: RPCRequest -> FloorM Value
handleRequest (RPCRequest method maybeParams) =
    let params = fromMaybe (object []) maybeParams
    in case method of
        "get_info" -> handleGetInfo
        "add_agent" -> handleAddAgent params
        "create_task" -> handleCreateTask params
        "process_code" -> handleProcessCode params
        _ -> return $ object ["status" .= ("error" :: Text), "message" .= ("Unknown method: " <> method)]

processRequest :: Text -> FloorM ()
processRequest line = do
    case JSON.eitherDecode (BSL.fromStrict $ TE.encodeUtf8 line) of
        Left err -> do
            let response = object ["status" .= ("error" :: Text), "message" .= ("Invalid JSON: " <> T.pack err)]
            liftIO $ BSL.putStrLn (JSON.encode response)
            liftIO $ hFlush stdout
        Right req -> do
            response <- handleRequest req
            liftIO $ BSL.putStrLn (JSON.encode response)
            liftIO $ hFlush stdout

mainLoop :: FloorM ()
mainLoop = forever $ do
    line <- liftIO TIO.getLine
    processRequest line

main :: IO ()
main = do
    hPutStrLn stderr "Haskell Department Floor (Floor 20) - Ready"
    hPutStrLn stderr "Domain: Pure functional programming, Type-safe systems, Compilers"
    hPutStrLn stderr "Architectural Law: Type safety > runtime checks, No partial functions"
    hPutStrLn stderr "Offices: Architecture, Implementation, Review, Test, Security, Manager"
    hFlush stderr
    
    evalStateT mainLoop emptyFloorState

File: floors/swift/DepartmentFloor.swift
#!/usr/bin/env swift
/**
 * FLOOR 13 - SWIFT JURISDICTION
 * Department Floor Implementation
 * 
 * Domain: Apple platforms (iOS, macOS, watchOS, tvOS), Server-side Swift, System programming
 * Architectural Law: Value semantics, protocol-oriented, memory safety via ARC
 */

import Foundation
import Dispatch

struct FloorAgent: Codable {
    let agentId: String
    let name: String
    let role: String
    let capabilities: [String]
}

enum TaskStatus: String, Codable {
    case pending
    case inProgress = "in_progress"
    case completed
    case failed
}

struct Task: Codable {
    let taskId: String
    let title: String
    var status: TaskStatus
    let assignedTo: String
    let createdAt: String
    var completedAt: String?
}

struct CodeAnalysis: Codable {
    let lines: Int
    let functions: Int
    let classes: Int
    let structs: Int
    let protocols: Int
    let language: String
    let memorySafetyScore: Double
    let protocolUsage: Int
}

struct JsonRpcRequest: Codable {
    let method: String
    let params: [String: AnyCodable]?
    let id: String?
}

struct JsonRpcResponse: Codable {
    let result: [String: AnyCodable]?
    let error: JsonRpcError?
    let id: String?
}

struct JsonRpcError: Codable {
    let code: Int
    let message: String
}

struct AnyCodable: Codable {
    let value: Any
    
    init(_ value: Any) {
        self.value = value
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if let intValue = try? container.decode(Int.self) {
            value = intValue
        } else if let doubleValue = try? container.decode(Double.self) {
            value = doubleValue
        } else if let stringValue = try? container.decode(String.self) {
            value = stringValue
        } else if let boolValue = try? container.decode(Bool.self) {
            value = boolValue
        } else if let arrayValue = try? container.decode([AnyCodable].self) {
            value = arrayValue.map { $0.value }
        } else if let dictValue = try? container.decode([String: AnyCodable].self) {
            value = dictValue.mapValues { $0.value }
        } else {
            value = NSNull()
        }
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        
        switch value {
        case let intValue as Int:
            try container.encode(intValue)
        case let doubleValue as Double:
            try container.encode(doubleValue)
        case let stringValue as String:
            try container.encode(stringValue)
        case let boolValue as Bool:
            try container.encode(boolValue)
        case let arrayValue as [Any]:
            try container.encode(arrayValue.map { AnyCodable($0) })
        case let dictValue as [String: Any]:
            try container.encode(dictValue.mapValues { AnyCodable($0) })
        default:
            try container.encodeNil()
        }
    }
}

enum Office {
    case architecture
    case implementation
    case review
    case test
    case security
    case manager
    
    var name: String {
        switch self {
        case .architecture: return "Architecture Office"
        case .implementation: return "Implementation Office"
        case .review: return "Review Office"
        case .test: return "Test Office"
        case .security: return "Security Office"
        case .manager: return "Manager Office"
        }
    }
    
    var jurisdiction: String {
        switch self {
        case .architecture: return "Protocol-oriented design, SwiftUI architecture, system design"
        case .implementation: return "Value types, protocol extensions, generics implementation"
        case .review: return "Code review, SwiftLint compliance, API guidelines"
        case .test: return "XCTest, UI testing, performance testing"
        case .security: return "Memory safety, data protection, keychain, secure coding"
        case .manager: return "SPM coordination, build management, release pipeline"
        }
    }
}

protocol Agent {
    var id: String { get }
    var office: Office { get }
    func execute(task: Task, completion: @escaping (Result<String, Error>) -> Void)
}

class ServiceAgent: Agent {
    let id: String
    let office: Office
    
    init(id: String, office: Office) {
        self.id = id
        self.office = office
    }
    
    func execute(task: Task, completion: @escaping (Result<String, Error>) -> Void) {
        DispatchQueue.global().asyncAfter(deadline: .now() + 0.01) {
            completion(.success("ServiceAgent \(self.id) completed task: \(task.title)"))
        }
    }
}

class DataModelAgent: Agent {
    let id: String
    let office: Office
    
    init(id: String, office: Office) {
        self.id = id
        self.office = office
    }
    
    func execute(task: Task, completion: @escaping (Result<String, Error>) -> Void) {
        DispatchQueue.global().asyncAfter(deadline: .now() + 0.01) {
            completion(.success("DataModelAgent \(self.id) validated value type: \(task.title)"))
        }
    }
}

class OperationsAgent: Agent {
    let id: String
    let office: Office
    
    init(id: String, office: Office) {
        self.id = id
        self.office = office
    }
    
    func execute(task: Task, completion: @escaping (Result<String, Error>) -> Void) {
        DispatchQueue.global().asyncAfter(deadline: .now() + 0.01) {
            completion(.success("OperationsAgent \(self.id) executed: \(task.title)"))
        }
    }
}

class SwiftDepartmentFloor {
    private let floorNumber = 13
    private let language = "swift"
    private let domain = "Apple platforms (iOS, macOS, watchOS, tvOS), Server-side Swift"
    
    private var agents: [String: FloorAgent] = [:]
    private var tasks: [String: Task] = [:]
    private var activeAgents: [String: Agent] = [:]
    
    private let offices: [Office] = [
        .architecture,
        .implementation,
        .review,
        .test,
        .security,
        .manager
    ]
    
    private let queue = DispatchQueue(label: "com.office.floor13", attributes: .concurrent)
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()
    
    init() {
        encoder.outputFormatting = []
        initializeAgents()
    }
    
    private func initializeAgents() {
        activeAgents["service_001"] = ServiceAgent(id: "service_001", office: .architecture)
        activeAgents["data_001"] = DataModelAgent(id: "data_001", office: .implementation)
        activeAgents["ops_001"] = OperationsAgent(id: "ops_001", office: .review)
    }
    
    func addAgent(agentId: String, name: String, role: String, capabilities: [String]) -> [String: Any] {
        let agent = FloorAgent(agentId: agentId, name: name, role: role, capabilities: capabilities)
        queue.async(flags: .barrier) {
            self.agents[agentId] = agent
        }
        return [
            "status": "success",
            "agent": [
                "agent_id": agent.agentId,
                "name": agent.name,
                "role": agent.role,
                "capabilities": agent.capabilities
            ]
        ]
    }
    
    func createTask(taskId: String, title: String, assignedTo: String) -> [String: Any] {
        let semaphore = DispatchSemaphore(value: 0)
        var task = Task(
            taskId: taskId,
            title: title,
            status: .pending,
            assignedTo: assignedTo,
            createdAt: ISO8601DateFormatter().string(from: Date())
        )
        
        if let agent = activeAgents[assignedTo] {
            task.status = .inProgress
            agent.execute(task: task) { result in
                switch result {
                case .success:
                    task.status = .completed
                    task.completedAt = ISO8601DateFormatter().string(from: Date())
                case .failure:
                    task.status = .failed
                }
                semaphore.signal()
            }
            _ = semaphore.wait(timeout: .now() + 5)
        }
        
        queue.async(flags: .barrier) {
            self.tasks[taskId] = task
        }
        
        return [
            "status": "success",
            "task": taskToDict(task)
        ]
    }
    
    func getFloorInfo() -> [String: Any] {
        var agentList: [[String: Any]] = []
        var taskList: [[String: Any]] = []
        
        queue.sync {
            agentList = agents.values.map { agent in
                [
                    "agent_id": agent.agentId,
                    "name": agent.name,
                    "role": agent.role,
                    "capabilities": agent.capabilities
                ]
            }
            taskList = tasks.values.map(taskToDict)
        }
        
        return [
            "floor_number": floorNumber,
            "language": language,
            "domain": domain,
            "offices": offices.map { ["name": $0.name, "jurisdiction": $0.jurisdiction] },
            "agent_count": agents.count,
            "task_count": tasks.count,
            "active_agents": activeAgents.count,
            "agents": agentList,
            "tasks": taskList,
            "features": [
                "Automatic Reference Counting (ARC)",
                "Value semantics by default",
                "Protocol-oriented programming",
                "Optional types for null safety",
                "Generics and associated types",
                "SwiftUI for declarative UI",
                "Swift Package Manager"
            ]
        ]
    }
    
    func processCode(code: String, operation: String) -> [String: Any] {
        switch operation {
        case "analyze":
            return analyzeCode(code)
        case "format":
            return formatCode(code)
        case "lint":
            return lintCode(code)
        case "compile":
            return compileCheck(code)
        default:
            return [
                "status": "error",
                "message": "Unknown operation: \(operation)"
            ]
        }
    }
    
    private func analyzeCode(_ code: String) -> [String: Any] {
        let lines = code.components(separatedBy: .newlines).count
        let functions = code.components(separatedBy: "func ").count - 1
        let classes = code.components(separatedBy: "class ").count - 1
        let structs = code.components(separatedBy: "struct ").count - 1
        let protocols = code.components(separatedBy: "protocol ").count - 1
        let memorySafetyScore = calculateMemorySafety(code)
        let protocolUsage = code.components(separatedBy: ": ").count - 1
        
        return [
            "status": "success",
            "analysis": [
                "lines": lines,
                "functions": functions,
                "classes": classes,
                "structs": structs,
                "protocols": protocols,
                "language": language,
                "memory_safety_score": memorySafetyScore,
                "protocol_usage": protocolUsage
            ]
        ]
    }
    
    private func calculateMemorySafety(_ code: String) -> Double {
        var score = 1.0
        
        if code.contains("UnsafeMutablePointer") || code.contains("UnsafeRawPointer") {
            score -= 0.3
        }
        if code.contains("force unwrap") || code.contains("!") {
            score -= 0.2
        }
        if !code.contains("weak") && !code.contains("unowned") && code.contains("class") {
            score -= 0.1
        }
        
        return max(0.0, score)
    }
    
    private func formatCode(_ code: String) -> [String: Any] {
        return [
            "status": "success",
            "formatted": true,
            "message": "Swift code formatted (SwiftFormat standard)",
            "changes": [
                "Applied value semantics",
                "Converted to protocol extensions",
                "Optimized collection operations"
            ]
        ]
    }
    
    private func lintCode(_ code: String) -> [String: Any] {
        var issues: [String] = []
        
        if !code.contains("import Foundation") && !code.contains("import ") {
            issues.append("Consider adding import statements")
        }
        
        if code.contains("!") && !code.contains("guard") {
            issues.append("Force unwrap (!) detected - prefer optional binding or guard")
        }
        
        if code.contains("class") && !code.contains("final") {
            issues.append("Consider marking classes as 'final' if not designed for inheritance")
        }
        
        if code.contains("var") && !code.contains("let") {
            issues.append("Prefer immutable 'let' over mutable 'var'")
        }
        
        return [
            "status": "success",
            "issues": issues,
            "severity": issues.isEmpty ? "none" : "warning"
        ]
    }
    
    private func compileCheck(_ code: String) -> [String: Any] {
        return [
            "status": "success",
            "compiled": true,
            "message": "Swift compilation check passed",
            "target": "Swift 5.9+"
        ]
    }
    
    private func taskToDict(_ task: Task) -> [String: Any] {
        var dict: [String: Any] = [
            "task_id": task.taskId,
            "title": task.title,
            "status": task.status.rawValue,
            "assigned_to": task.assignedTo,
            "created_at": task.createdAt
        ]
        if let completedAt = task.completedAt {
            dict["completed_at"] = completedAt
        }
        return dict
    }
    
    func handleRequest(_ requestString: String) -> String {
        guard let data = requestString.data(using: .utf8) else {
            return encodeError(code: -32700, message: "Parse error")
        }
        
        do {
            let request = try decoder.decode(JsonRpcRequest.self, from: data)
            let result = try executeMethod(request)
            
            let response = JsonRpcResponse(
                result: result.mapValues { AnyCodable($0) },
                error: nil,
                id: request.id
            )
            
            let responseData = try encoder.encode(response)
            return String(data: responseData, encoding: .utf8) ?? ""
        } catch {
            return encodeError(code: -32603, message: "Internal error: \(error.localizedDescription)")
        }
    }
    
    private func executeMethod(_ request: JsonRpcRequest) throws -> [String: Any] {
        switch request.method {
        case "get_info":
            return getFloorInfo()
        case "add_agent":
            guard let params = request.params else {
                throw NSError(domain: "Floor13", code: -1, userInfo: [NSLocalizedDescriptionKey: "Missing parameters"])
            }
            return addAgent(
                agentId: params["agent_id"]?.value as? String ?? "",
                name: params["name"]?.value as? String ?? "",
                role: params["role"]?.value as? String ?? "",
                capabilities: params["capabilities"]?.value as? [String] ?? []
            )
        case "create_task":
            guard let params = request.params else {
                throw NSError(domain: "Floor13", code: -1, userInfo: [NSLocalizedDescriptionKey: "Missing parameters"])
            }
            return createTask(
                taskId: params["task_id"]?.value as? String ?? "",
                title: params["title"]?.value as? String ?? "",
                assignedTo: params["assigned_to"]?.value as? String ?? ""
            )
        case "process_code":
            guard let params = request.params else {
                throw NSError(domain: "Floor13", code: -1, userInfo: [NSLocalizedDescriptionKey: "Missing parameters"])
            }
            return processCode(
                code: params["code"]?.value as? String ?? "",
                operation: params["operation"]?.value as? String ?? ""
            )
        default:
            return [
                "status": "error",
                "message": "Unknown method: \(request.method)"
            ]
        }
    }
    
    private func encodeError(code: Int, message: String) -> String {
        let response = JsonRpcResponse(
            result: nil,
            error: JsonRpcError(code: code, message: message),
            id: nil
        )
        
        if let data = try? encoder.encode(response),
           let string = String(data: data, encoding: .utf8) {
            return string
        }
        return "{\"error\":{\"code\":\(code),\"message\":\"\(message)\"}}"
    }
}

let floor = SwiftDepartmentFloor()

while let line = readLine() {
    guard !line.isEmpty else { continue }
    let response = floor.handleRequest(line)
    print(response)
    fflush(stdout)
}
